src/rpcq.lisp

```diff
@@ -573,38 +282,95 @@ We distinguish between the following options for any field type:
-                                           (,(accessor slot) ,obj)))))))))))
+    (alexandria:with-gensyms (obj init-args)
+      (let ((slot-names (mapcar #'car field-specs)))
+        `(progn
+           (defclass ,class-name ,parent-name
+             ,(mapcar #'make-slot-spec field-specs)
+             ,@(when documentation
+                 `((:documentation ,(format-documentation-string documentation)))))
+           
+           (defmethod shared-initialize ((instance ,class-name) slot-names
+                                           &rest initargs
+                                           &key &allow-other-keys)
+               ,@(mapcan #'slot-initialization-deprecation-warnings field-specs)
+               (call-next-method))
+           
+           ,@(mapcan #'slot-accessor-deprecation-warnings field-specs)
+           
+           (defmethod %serialize ((,obj ,class-name))
+             (%%serialize ,obj (make-hash-table :test #'equal)))
+           
+           (defmethod %%serialize ((,obj ,class-name) (hash-table hash-table))
+             (with-slots ,slot-names ,obj
+               ,@(loop :for slot :in slot-names
+                       :collect `(setf (gethash ,(symbol-name slot) hash-table)
+                                       (%serialize ,slot)))
+               (call-next-method)
+               (setf (gethash "_type" hash-table) ,(symbol-name class-name))
+               hash-table))
+
+           (defmethod %deserialize-struct ((type (eql ',class-name)) (payload hash-table))
+             (assert (string= (gethash "_type" payload) ,(symbol-name class-name)))
+             (let ((,init-args (%%deserialize-struct type payload)))
+               (apply 'make-instance
+                      ',class-name
+                      ,init-args)))
+           
+           (defmethod %%deserialize-struct ((type (eql ',class-name)) (payload hash-table))
+             ,(cond
+                (parent-name
+                 `(nconc
+                   (list ,@(mapcan (init-spec 'payload) slot-names))
+                   (%%deserialize-struct ',@parent-name payload)))
+                (t
+                 `(list ,@(mapcan (init-spec 'payload) slot-names)))))
+
+           (defmethod print-object ((,obj ,class-name) stream)
+             (print-unreadable-object (,obj stream :type t)
+               (pprint-indent :block 2)
+               (%print-slots ,obj stream)))
+           
+           (defmethod %print-slots ((,obj ,class-name) stream)
+             ,@(loop :for spec :in field-specs
+                     :unless (getf (cdr spec) :deprecated-by)
+                       :collect `(format stream
+                                         "~&  ~A -> ~S"
+                                         ,(symbol-name (car spec))
+                                         (,(accessor (car spec)) ,obj)))
+             (call-next-method))
+           
+           nil)))))
+
+(defmethod %print-slots (obj stream)
```

Comment:
Do we want these? It seems dangerous to return NIL in general.

Reply:
I'm happy to take CLOS advice here: I don't want `call-next-method` to vomit, and RPCQ message definition classes in Lisp currently don't inherit from any kind of unified parent class. Maybe I can be more clever about my `defmethod` declaration, or maybe I should install an (abstract) superclass. What's established praxis?